internal/cluster/validator.go

```diff
@@ -0,0 +1,254 @@
+	return boolValue(err == nil)
+}
+
+func (v *clusterValidator) printIsApiVipValid(context *clusterValidationContext, status validationStatus) string {
+	switch status {
+	case ValidationPending:
+		return "Machine network CIDR or API vip or Ingress vip is undefined"
+	case ValidationSuccess:
+		return fmt.Sprintf("%s %s belongs to machine CIDR and not in use ", ApiVipName, context.cluster.APIVip)
+	case ValidationFailure:
+		return fmt.Sprintf("%s %s does not belong to machine CIDR or already in use ", ApiVipName, context.cluster.APIVip)
+	default:
+		return fmt.Sprintf("Unexpected status %s", status)
+	}
+}
+
+func (v *clusterValidator) isIngressVipDefined(c *clusterValidationContext) validationStatus {
+	return boolValue(c.cluster.IngressVip != "")
+}
+
+func (v *clusterValidator) printIsIngressVipDefined(context *clusterValidationContext, status validationStatus) string {
+	switch status {
+	case ValidationFailure:
+		return "Ingress vip is undefined"
+	case ValidationSuccess:
+		return "Ingress vip is defined"
+	default:
+		return fmt.Sprintf("Unexpected status %s", status)
+	}
+}
+func (v *clusterValidator) isIngressVipValid(c *clusterValidationContext) validationStatus {
+	if c.cluster.MachineNetworkCidr == "" || c.cluster.APIVip == "" || c.cluster.IngressVip == "" {
+		return ValidationPending
+	}
+	err := network.VerifyVip(c.cluster.Hosts, c.cluster.MachineNetworkCidr, c.cluster.IngressVip, IngressVipName,
+		true, v.log)
+	return boolValue(err == nil)
+}
+
+func (v *clusterValidator) printIsIngressVipValid(context *clusterValidationContext, status validationStatus) string {
+	switch status {
+	case ValidationPending:
+		return "Machine network CIDR or API vip or Ingress vip is undefined"
+	case ValidationSuccess:
+		return fmt.Sprintf("%s %s belongs to machine CIDR and not in use ", IngressVipName, context.cluster.IngressVip)
+	case ValidationFailure:
+		return fmt.Sprintf("%s %s does not belong to machine CIDR or already in use ", IngressVipName, context.cluster.IngressVip)
+	default:
+		return fmt.Sprintf("Unexpected status %s", status)
+	}
+}
+
+func (v *clusterValidator) hasExactlyThreeKnownMasters(c *clusterValidationContext) validationStatus {
+	mappedMastersByRole := MapMasterHostsByStatus(c.cluster)
+
+	mastersInKnown, ok := mappedMastersByRole[models.HostStatusKnown]
+	return boolValue(ok && len(mastersInKnown) == MinHostsNeededForInstallation)
+}
+
+func (v *clusterValidator) printHasExactlyThreeKnownMasters(context *clusterValidationContext, status validationStatus) string {
+	switch status {
+	case ValidationSuccess:
+		return fmt.Sprintf("Cluster has %d known master hosts", MinHostsNeededForInstallation)
+	case ValidationFailure:
+		return fmt.Sprintf("cluster is expected to have exactly %d known master to be installed", MinHostsNeededForInstallation)
+	default:
+		return fmt.Sprintf("Unexpected status %s", status)
+	}
+}
+
+func (v *clusterValidator) noHostInPendingForInput(c *clusterValidationContext) validationStatus {
```

Comment:
Why is this validation needed?

Reply:
When we reviewed the design document we agreed that if there is one host in pending-for-input because no role was defined - the cluster should be in pending-for-input. 
If one of the hosts is in insufficient status - the validation that all hosts are ready to install will be false.
@avishayt what do you think?

Reply:
I understand, I think it complicates things. Its removal will make things clearer

Reply:
cluster doesn't have pending-for-input state currently.  anyway soon with auto-assigned roles this won't happen anymore.

Reply:
when we will have the auto-assigned feature - the cluster will be in pending for input if the machine CIDR, API VIP or Ingress VIP are not defined.